sdks/v4-sdk/src/entities/pool.ts

```diff
@@ -0,0 +1,309 @@
+import invariant from 'tiny-invariant'
+import { keccak256 } from '@ethersproject/solidity'
+import { BigintIsh, Currency, CurrencyAmount, Price } from '@uniswap/sdk-core'
+import {
+  v3Swap,
+  NoTickDataProvider,
+  Tick,
+  TickConstructorArgs,
+  TickDataProvider,
+  TickListDataProvider,
+  TickMath,
+} from '@uniswap/v3-sdk'
+import { defaultAbiCoder, isAddress } from 'ethers/lib/utils'
+import { sortsBefore } from '../utils/sortsBefore'
+import { ADDRESS_ZERO, NEGATIVE_ONE, Q192 } from '../utils/internalConstants'
+import JSBI from 'jsbi'
+
+const NO_TICK_DATA_PROVIDER_DEFAULT = new NoTickDataProvider()
+
+export type PoolKey = {
+  currency0: string
+  currency1: string
+  fee: number
+  tickSpacing: number
+  hooks: string
+}
+
+/**
+ * Represents a V4 pool
+ */
+export class Pool {
+  public readonly currency0: Currency
+  public readonly currency1: Currency
+  public readonly fee: number
+  public readonly tickSpacing: number
+  public readonly sqrtRatioX96: JSBI
+  public readonly hooks: string // address
+  public readonly liquidity: JSBI
+  public readonly tickCurrent: number
+  public readonly tickDataProvider: TickDataProvider
+
+  // backwards compatibility with Pool interface v2/v3 sdk
+  public readonly token0: Currency
```

Comment:
Consider using getters to make these type aliases without having fully separate instances -- avoids potential for mis-setting it, i.e.

```
public get token0(): Currency {
   return this.currency0;
}
```